### PR TITLE
refactor: complex voucher tools

### LINKED ISSUE #1956: fix(journal voucher) : Automated description for prise en charge
Add an automated description for the prise en charge tool.

When the prise en charge tool is used, the description field in the journal voucher is not filled, the user is obliged to fill it manually.

This issue is about to fill the description field automatically with meaningful description when the prise en charge tool is used.

### PR DESCRIPTION
**Pull request by jniles**
This PR tracks work towards a large-scale refactor of Voucher Tools.  Each Voucher Tool is being revisited and rewritten to increase performance and address any reported bugs.  Duplicated code is being consolidated into shared services.

~~**Note** - This work is still ongoing and should not be merged yet.~~  This can now be merged anytime.

**Major Changes**
 1. Replace bhBreadcrumb with a simple headercrumb.  This change reduces the JS code by about 25 lines.  It also makes the voucher tools much easier to update, re-order, and change.  See #1923.
 2. Move shared code to the `VoucherToolKitService`.  This is particular evident with the `generateRow()` function that was used identically in ever controller.
 3. All forms that require accounts now use the `bhAccountSelect` component instead of manually downloading accounts themselves.
 4. Update Prise En Charge text.  Fixes #1956.
 5. Ensure modal ui-grids function correctly and more efficiently.
   a. Cost/balance sorting is now correct.
   b. Remove unnecessary cell templates (e.g. `balance.cell.html`)
   c. Consolidate `onSelectionChanged` listeners - reduces number of listeners by 50%.
   d. Consolidate aggregation functions into `aggregate()`.  No longer defines a new function on every selection change event.

-----
Thank you for contributing!

Before submitting this pull request, please verify that you have:
 - [x] Run your code through eslint.
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the online review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

File: client/src/modules/vouchers/complex-voucher.ctrl.js [hunk 1]
```diff
@@ -76,32 +61,31 @@ function ComplexJournalVoucherController(Vouchers, $translate, Currencies, Sessi
     vm.gridApi = api;
   }
 
-  /** ======================== voucher tools ========================== */
+  vm.openConventionPaymentModal = function openConventionPaymentModal() {
+    Toolkit.openConventionPaymentModal()
+      .then(processVoucherToolRows);
+  };
```
Comment (mbayopanda) [inline client/src/modules/vouchers/complex-voucher.ctrl.js:67]: 👍

File: client/src/modules/vouchers/complex-voucher.ctrl.js [hunk 2]
```diff
@@ -45,22 +46,6 @@ function ComplexJournalVoucherController(Vouchers, $translate, Currencies, Sessi
...
-  }];
-
-  // breadcrumb dropdown
-  vm.dropdown = [{
-    label  : 'FORM.LABELS.MENU',
-    color  : 'btn-default',
-    icon   : 'fa-cogs',
-    option : Toolkit.options,
-  }];
-
```
Comment (mbayopanda) [inline client/src/modules/vouchers/complex-voucher.ctrl.js:63]: 💃 the mystery of the black box of `breadcrumb component` dynamic tools is now exposed in a clear way.

File: client/src/modules/vouchers/toolkit/cash_transfer/cash_transfer.html [hunk 3]
```diff
@@ -37,41 +37,19 @@
...
-        <span class="input-group-addon">{{ ToolCtrl.cashbox.symbol }}</span>
-        <input class="form-control" type="number" ng-model="ToolCtrl.amount" ng-disabled="!ToolCtrl.cashbox" required>
-      </div>
-    </div>
+    <bh-account-select
+      label="VOUCHERS.GLOBAL.TRANSFER_ACCOUNT"
+      on-select-callback="ToolCtrl.onSelectAccountCallback(account)"
+      exclude-title-accounts="true">
+      (<label class="control-label" translate>FORM.LABELS.CREDIT</label>)
+    </bh-account-select>
```
Comment (mbayopanda) [inline client/src/modules/vouchers/toolkit/cash_transfer/cash_transfer.html:46]: 👍 Ooura... Component

**Review by mbayopanda — COMMENTED** [review]
The code LGTM

File: client/src/modules/vouchers/toolkit/cash_transfer/cash_transfer.html (hunk 3 above)
Comment (jniles) [reply client/src/modules/vouchers/toolkit/cash_transfer/cash_transfer.html:46]: [Ooura](http://www.dictionary.com/misspelling?term=ooura&s=t) :laughing:

**Comment by jniles** [comment]
Ugh.  I'm not sure if it is here or in `master`, but the voucher type selection is broken.  I'll try to fix it here.

**Comment by jniles** [comment]
Discovered that the voucher type selection was broken on ui-select versions greater than 0.17.1.  This is fixed in https://github.com/IMA-WorldHealth/bhima-2.X/pull/2078/commits/1c42da2ca2f58b937f232a6069f202da6ef4a16a.

**Comment by jniles** [comment]
Yay!  This is due to a real bug! 🐛

**Comment by jniles** [comment]
bors r=mbayopanda

**Comment by jniles** [comment]
bors r=mbayopanda

... try again.

**Pull request merged**